test/tls-provider.c

```diff
@@ -0,0 +1,453 @@
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/core_names.h>
+#include <openssl/core_numbers.h>
+#include <openssl/rand.h>
+#include <openssl/params.h>
+/* For TLS1_3_VERSION */
+#include <openssl/ssl.h>
+
+int tls_provider_init(const OSSL_CORE_HANDLE *handle,
+                      const OSSL_DISPATCH *in,
+                      const OSSL_DISPATCH **out,
+                      void **provctx);
+
+#define XOR_KEY_SIZE 32
+
+/*
+ * Top secret. This algorithm only works if no one knows what this number is.
+ * Please don't tell anyone what it is.
+ */
+const unsigned char private_constant[XOR_KEY_SIZE] = {
+    0xd3, 0x6b, 0x54, 0xec, 0x5b, 0xac, 0x89, 0x96, 0x8c, 0x2c, 0x66, 0xa5,
+    0x67, 0x0d, 0xe3, 0xdd, 0x43, 0x69, 0xbc, 0x83, 0x3d, 0x60, 0xc7, 0xb8,
+    0x2b, 0x1c, 0x5a, 0xfd, 0xb5, 0xcd, 0xd0, 0xf8
+};
+
+
+typedef struct xorkey_st {
+    unsigned char privkey[XOR_KEY_SIZE];
+    unsigned char pubkey[XOR_KEY_SIZE];
+    int hasprivkey;
+    int haspubkey;
+} XORKEY;
+
+/* We define a dummy TLS group called "xorgroup" for test purposes */
+
+unsigned int group_id = 65137; /* IANA reserved for private use */
+unsigned int secbits = 128;
+unsigned int mintls = TLS1_3_VERSION;
+unsigned int maxtls = 0;
+unsigned int mindtls = -1;
+unsigned int maxdtls = -1;
+
+#define GROUP_NAME "xorgroup"
+#define GROUP_NAME_INTERNAL "xorgroup-int"
+#define ALGORITHM "XOR"
+
+static const OSSL_PARAM xor_group_params[] = {
+    OSSL_PARAM_utf8_string(OSSL_CAPABILITY_TLS_GROUP_NAME,
+                           GROUP_NAME, sizeof(GROUP_NAME)),
+    OSSL_PARAM_utf8_string(OSSL_CAPABILITY_TLS_GROUP_NAME_INTERNAL,
+                           GROUP_NAME_INTERNAL, sizeof(GROUP_NAME_INTERNAL)),
+    OSSL_PARAM_utf8_string(OSSL_CAPABILITY_TLS_GROUP_ALG, ALGORITHM,
+                           sizeof(ALGORITHM)),
+    OSSL_PARAM_uint(OSSL_CAPABILITY_TLS_GROUP_ID, &group_id),
+    OSSL_PARAM_uint(OSSL_CAPABILITY_TLS_GROUP_SECURITY_BITS, &secbits),
+    OSSL_PARAM_int(OSSL_CAPABILITY_TLS_GROUP_MIN_TLS, &mintls),
+    OSSL_PARAM_int(OSSL_CAPABILITY_TLS_GROUP_MAX_TLS, &maxtls),
+    OSSL_PARAM_int(OSSL_CAPABILITY_TLS_GROUP_MIN_DTLS, &mindtls),
+    OSSL_PARAM_int(OSSL_CAPABILITY_TLS_GROUP_MAX_DTLS, &maxdtls),
+    OSSL_PARAM_END
+};
+
+static int tls_prov_get_capabilities(void *provctx, const char *capability,
+                                     OSSL_CALLBACK *cb, void *arg)
+{
+    /* We're only adding one group so we only call the callback once */
+    if (strcmp(capability, "TLS-GROUP") == 0)
+        return cb(xor_group_params, arg);
+
+    /* We don't support this capability */
+    return 0;
+}
+
+/*
+ * Dummy "XOR" Key Exchange algorithm. We just xor the private and public keys
+ * together. Don't use this!
```

Comment:
Cant wait to see this in production code :)

Reply:
It's a step up from the NULL algorithm.

Reply:
Clearly we need to randomize what `group_id` we use (at runtime) so that it won't interoperate.

Reply:
Added randmisation of the group_id